src/psbt/mod.rs

```diff
@@ -366,32 +371,408 @@ fn sanity_check(psbt: &Psbt) -> Result<(), Error> {
+    /// set to [`None`] while computing sighash for pre-taproot outputs.
+    /// The function also updates the sighash cache with transaction computed during sighash computation of this input
+    ///
+    /// # Arguments:
+    ///
+    /// * `idx`: The input index of psbt to sign
+    /// * `cache`: The [`sighash::SigHashCache`] for used to cache/read previously cached computations
+    /// * `tapleaf_hash`: If the output is taproot, compute the sighash for this particular leaf.
+    fn sighash_msg<T: Deref<Target = bitcoin::Transaction>>(
+        &self,
+        idx: usize,
+        cache: &mut SigHashCache<T>,
+        tapleaf_hash: Option<TapLeafHash>,
+    ) -> Result<PsbtSigHashMsg, SigHashError>;
+}
+
+impl PbstOps for Psbt {
+    /// See the triat level documentation for details
+    fn finalize<C: secp256k1::Verification>(
+        &mut self,
+        secp: &secp256k1::Secp256k1<C>,
+    ) -> Result<(), Error> {
+        finalizer::finalize_helper(self, secp, /*allow_mall*/ false)
+    }
+
+    /// See the triat level documentation for details
+    fn finalize_mall<C: secp256k1::Verification>(
+        &mut self,
+        secp: &secp256k1::Secp256k1<C>,
+    ) -> Result<(), Error> {
+        finalizer::finalize_helper(self, secp, /*allow_mall*/ true)
+    }
+
+    /// See the triat level documentation for details
+    fn extract<C: secp256k1::Verification>(
+        &self,
+        secp: &Secp256k1<C>,
+    ) -> Result<bitcoin::Transaction, Error> {
+        sanity_check(self)?;
+
+        let mut ret = self.unsigned_tx.clone();
+        for (n, input) in self.inputs.iter().enumerate() {
+            if input.final_script_sig.is_none() && input.final_script_witness.is_none() {
+                return Err(Error::InputError(InputError::MissingWitness, n));
+            }
+
+            if let Some(witness) = input.final_script_witness.as_ref() {
+                ret.input[n].witness = witness.clone();
+            }
+            if let Some(script_sig) = input.final_script_sig.as_ref() {
+                ret.input[n].script_sig = script_sig.clone();
+            }
+        }
+        interpreter_check(self, secp)?;
+        Ok(ret)
+    }
+
+    /// See the triat level documentation for details
+    fn update_desc(
+        &mut self,
+        input_index: usize,
+        desc: &Descriptor<DescriptorPublicKey>,
+        range: Range<u32>,
+    ) -> Result<(), UxtoUpdateError> {
+        if input_index >= self.inputs.len() {
+            return Err(UxtoUpdateError::IndexOutOfBounds(
+                input_index,
+                self.inputs.len(),
+            ));
+        }
+        let mut derived_desc = None;
+        {
+            let inp_spk = finalizer::get_scriptpubkey(self, input_index)
+                .map_err(|_e| UxtoUpdateError::MissingInputUxto)?;
+            let secp = secp256k1::Secp256k1::verification_only();
+
+            if !(range.start < range.end) {
```

Comment:
In 27679de0b4ba07da7e007b1bf046834ebf0abae2:

I'm confused about the purpose of this if-clause. If the user provides an empty range shouldn't we just do nothing?

Reply:
Ohh, is the issue that if you try to iterate on a bad range it'll run (effectively) forever? I think you should comment this, and also change the if condition to `if range.start > range.end` and also I continue to think this should be empty, rather than deriving the 0 index. Users who explicitly want 0 can provide `0..0`.

Reply:
If the descriptor is already is derived and contains no wildcards, we still need to process it. Will add a comment here.

Reply:
I wanted to use the `range.is_empty()` API, but that some weird behavior in rust 1.29. Specifically, I can't select candidate 1 amongst the following:

```
511 |         if Range::<u32>::is_empty(&range) { // required for MSRV
    |            ^^^^^^^^^^^^^^^^^^^^^^ multiple `is_empty` found
    |
    = note: candidate #1 is defined in an impl for the type `std::ops::Range<_>`
    = note: candidate #2 is defined in an impl of the trait `std::iter::ExactSizeIterator`
```

I have tried a lot of ways to do this and rust IRC/discord could not help here.

Reply:
I think it might make more sense to take an `Option<Range>`?